README.md
# Fork of dhcpdump
Parse DHCP packets from a network interface

This fork of `dhcpdump` 1.8 by Edwin Groothuis, edwin@mavetju.org (http://www.mavetju.org) collects in a single place bug-fixes and improvements collected over the last 10+ years.

## Changes since 1.8

* Fix the DHCP flags calculation
* Print option 82 content in a more usable way
* Avoid OOB access for the undefined string values
* Get ethertype in edian agnostic way
* Add/remove headers
* Use `char` for strings and `uint8_t` for binary data
* Let the `Makefile` use environment variables
* Spelling fixes
* Remove the unused `strsep` implementation
* Use a stricter filer for DHCP packets
* Add an option to dump packet content in HEX
* More consistent alignment of output
* Avoid extra new lines in parameter request list (option 55)
* Print client identifier as HEX+ASCII
* Print mac addresses with leading zeroes
* Add support for VLAN tagged traffic



CHANGES
Old changes are kept for historical purpose. The latest changes
are in git history.


$Id: CHANGES,v 1.11 2004/10/31 11:22:58 mavetju Exp $

version 1.8
	- Use pcap instead of tcpdump output

version 1.7
	- Support for tcpdump 3.8.x

version 1.6
	- Option 82, submitted by Phil Pennock <pdp@nl.demon.net>
	- Add fflush(), submitted by Gary Colman <gary@colmans.net>
	- Fix problem with padding, they were not handled properly
	  and caused old memory to be reused.

version 1.5
	- better input checking regarding validity of the header
	  of the UDP packet (bug anton.holleman@nominum.com)
	- Added some DHCP options I found on the INIA website.
	  I can't parse all of them yet since I'm not sure how they
	  look like.

version 1.4
	- added auto-configure script
	- fixed problem with strsep() under Solaris
	- added FILES, CONTACT

version 1.3
	- added chaddr filter, contributed by Peter Apian-Bennewitz
	  <apian@ise.fhg.de>
	- client-identifier is now printed as long as it is.
	  (bug anton.holleman@nominum.com)

version 1.2
	- forgot to update the man-page.

version 1.1
	- Sometimes, specially with bootp-packets, the option-packet
	  would contain rubbish data which dhcpdump tried to parse.
	  And that caused an overflow in the arrays, which caused
	  a segfault.
	- Added more tables so the numerical values of fields (like
	  network type, requested parameter list) are not only
	  numeric anymore.
	- Better layout of hex-strings.

version 1.0
	- Initial release


strsep.c
//
// $Id: strsep.c,v 1.1 2002/01/26 13:19:04 mavetju Exp $
//
// strsep() for Solaris, which doesn't have it by default.
//

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <string.h>
#include <stdio.h>

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)strsep.c	8.1 (Berkeley) 6/4/93";
#endif /* LIBC_SCCS and not lint */
#ifndef lint
static const char rcsid[] =
  "$FreeBSD: src/lib/libc/string/strsep.c,v 1.2.12.1 2001/07/09 23:30:07 obrien Exp $";
#endif

/*
 * Get next token from string *stringp, where tokens are possibly-empty
 * strings separated by characters from delim.
 *
 * Writes NULs into the string at *stringp to end tokens.
 * delim need not remain constant from call to call.
 * On return, *stringp points past the last NUL written (if there might
 * be further tokens), or is NULL (if there are definitely no more tokens).
 *
 * If *stringp is NULL, strsep returns NULL.
 */
char *
strsep(stringp, delim)
	register char **stringp;
	register const char *delim;
{
	register char *s;
	register const char *spanp;
	register int c, sc;
	char *tok;

	if ((s = *stringp) == NULL)
		return (NULL);
	for (tok = s;;) {
		c = *s++;
		spanp = delim;
		do {
			if ((sc = *spanp++) == c) {
				if (c == 0)
					s = NULL;
				else
					s[-1] = 0;
				*stringp = s;
				return (tok);
			}
		} while (sc != 0);
	}
	/* NOTREACHED */
}


dhcp_options.h
//
// $Id: dhcp_options.h,v 1.4 2004/10/31 11:22:58 mavetju Exp $
//

//
// this list was stolen from The DHCP Handbook by Droms and Lemon, Appendix D
//

// The first comment is the number, the last parameter is if it's verbosed
const char *dhcp_options[] = {
/*   0 */ "pad",
/*   1 */ "Subnet mask",
/*   2 */ "Time offset",
/*   3 */ "Routers",
/*   4 */ "Time server",
/*   5 */ "Name server",
/*   6 */ "DNS server",
/*   7 */ "Log server",
/*   8 */ "Cookie server",
/*   9 */ "LPR server",
/*  10 */ "Impress server",
/*  11 */ "Resource location server",
/*  12 */ "Host name",
/*  13 */ "Boot file size",
/*  14 */ "Merit dump file",
/*  15 */ "Domainname",
/*  16 */ "Swap server",
/*  17 */ "Root path",
/*  18 */ "Extensions path",
/*  19 */ "IP forwarding",
/*  20 */ "Non-local source routing",
/*  21 */ "Policy filter",
/*  22 */ "Maximum datagram reassembly size",
/*  23 */ "Default IP TTL",
/*  24 */ "Path MTU aging timeout",
/*  25 */ "Path MTU plateau table",
/*  26 */ "Interface MTU",
/*  27 */ "All subnets local",
/*  28 */ "Broadcast address",
/*  29 */ "Perform mask discovery",
/*  30 */ "Mask supplier",
/*  31 */ "Perform router discovery",
/*  32 */ "Router solicitation",
/*  33 */ "Static route",
/*  34 */ "Trailer encapsulation",
/*  35 */ "ARP cache timeout",
/*  36 */ "Ethernet encapsulation",
/*  37 */ "TCP default TTL",
/*  38 */ "TCP keepalive interval",
/*  39 */ "TCP keepalive garbage",
/*  40 */ "NIS domain",
/*  41 */ "NIS servers",
/*  42 */ "NTP servers",
/*  43 */ "Vendor specific info",
/*  44 */ "NetBIOS name server",
/*  45 */ "NetBIOS datagram distribution server",
/*  46 */ "NetBIOS node type",
/*  47 */ "NetBIOS scope",
/*  48 */ "X Window System font server",
/*  49 */ "X Window System display server",
/*  50 */ "Request IP address",
/*  51 */ "IP address leasetime",
/*  52 */ "Option overload",
/*  53 */ "DHCP message type",
/*  54 */ "Server identifier",
/*  55 */ "Parameter Request List",
/*  56 */ "Message",
/*  57 */ "Maximum DHCP message size",
/*  58 */ "Renewal Time T1",
/*  59 */ "Rebinding Time T2",
/*  60 */ "Vendor class identifier",
/*  61 */ "Client-identifier",
/*  62 */ "Netware/IP domain name",
/*  63 */ "Netware/IP domain information",
/*  64 */ "NIS+ domain",
/*  65 */ "NIS+ servers",
/*  66 */ "TFTP server name",
/*  67 */ "Bootfile name",
/*  68 */ "Mobile IP home agent",
/*  69 */ "SMTP server",
/*  70 */ "POP3 server",
/*  71 */ "NNTP server",
/*  72 */ "WWW server",
/*  73 */ "Finger server",
/*  74 */ "IRC server",
/*  75 */ "StreetTalk server",
/*  76 */ "StreetTalk directory assistance server",
/*  77 */ "User-class Identification",
/*  78 */ "SLP-directory-agent",
/*  79 */ "SLP-service-scope",
/*  80 */ "Naming Authority",
/*  81 */ "Client FQDN",
/*  82 */ "Relay Agent Information",
/*  83 */ "Agent Remote ID",
/*  84 */ "Agent Subnet Mask",
/*  85 */ "NDS server",
/*  86 */ "NDS tree name",
/*  87 */ "NDS context",
/*  88 */ "IEEE 1003.1 POSIX",
/*  89 */ "FQDN",
/*  90 */ "Authentication",
/*  91 */ "Vines TCP/IP",
/*  92 */ "Server Selection",
/*  93 */ "Client System",
/*  94 */ "Client NDI",
/*  95 */ "LDAP",
/*  96 */ "IPv6 Transitions",
/*  97 */ "UUID/GUID",
/*  98 */ "UPA servers",
/*  99 */ "GEOCONF_CIVIC",
/* 100 */ "Printer Name",
/* 101 */ "MDHCP",
/* 102 */ "REMOVED/Unassigned",
/* 103 */ "REMOVED/Unassigned",
/* 104 */ "REMOVED/Unassigned",
/* 105 */ "REMOVED/Unassigned",
/* 106 */ "REMOVED/Unassigned",
/* 107 */ "REMOVED/Unassigned",
/* 108 */ "Swap Path",
/* 109 */ "DHCPv4 over DHCPv6 Source Address",
/* 110 */ "IPX Compatibility",
/* 111 */ "???",
/* 112 */ "Netinfo Address",
/* 113 */ "Netinfo Tag",
/* 114 */ "URL",
/* 115 */ "DHCP Failover",
/* 116 */ "DHCP Autoconfiguration",
/* 117 */ "Name Service Search",
/* 118 */ "Subnet selection",
/* 119 */ "Domain Search",
/* 120 */ "SIP Servers DHCP Option",
/* 121 */ "Classless Static Route",
/* 122 */ "CableLabs Client Configuration",
/* 123 */ "GeoConf Option",
/* 124 */ "V-I Vendor Class",
/* 125 */ "V-I Vendor-Specific Info",
/* 126 */ "REMOVED/Unassigned",
/* 127 */ "REMOVED/Unassigned",
/* 128 */ "PXE/Etherboot signature/DOCSIS",
/* 129 */ "PXE/Kernel options",
/* 130 */ "PXE/Ethernet interface",
/* 131 */ "PXE/Remote statistics server",
/* 132 */ "PXE/802.1Q VLAN ID",
/* 133 */ "PXE/802.1D/p Layer 2 Priority",
/* 134 */ "PXE/Diffserv Code Point for VoIP",
/* 135 */ "PXE/HTTP Proxy for phone",
/* 136 */ "OPTION_PANA_AGENT",
/* 137 */ "OPTION_V4_LOST",
/* 138 */ "OPTION_CAPWAP_AC_V4",
/* 139 */ "OPTION-IPv4_Address-MoS",
/* 140 */ "OPTION-IPv4_FQDN-MoS",
/* 141 */ "SIP UA Configuration SDomains",
/* 142 */ "OPTION-IPv4_Address-ANDSF",
/* 143 */ "OPTION_V4_SZTP_REDIRECT",
/* 144 */ "GeoLoc/HP - TFTP file",
/* 145 */ "FORCERENEW_NONCE_CAPABLE",
/* 146 */ "RDNSS Selection",
/* 147 */ "OPTION_V4_DOTS_RI",
/* 148 */ "OPTION_V4_DOTS_ADDRESS",
/* 149 */ "???",
/* 150 */ "TFTP server address",
/* 151 */ "status-code",
/* 152 */ "base-time",
/* 153 */ "start-time-of-state",
/* 154 */ "query-start-time",
/* 155 */ "query-end-time",
/* 156 */ "dhcp-state",
/* 157 */ "data-source",
/* 158 */ "OPTION_V4_PCP_SERVER",
/* 159 */ "OPTION_V4_PORTPARAMS",
/* 160 */ "???",
/* 161 */ "OPTION_MUD_URL_V4",
/* 162 */ "OPTION_V4_DNR",
/* 163 */ "???",
/* 164 */ "???",
/* 165 */ "???",
/* 166 */ "???",
/* 167 */ "???",
/* 168 */ "???",
/* 169 */ "???",
/* 170 */ "???",
/* 171 */ "???",
/* 172 */ "???",
/* 173 */ "???",
/* 174 */ "???",
/* 175 */ "Etherboot (Tentatively 2005-06-23)",
/* 176 */ "IP Telephone (Tentatively 2005-06-23)",
/* 177 */ "Etherboot (Tentatively 2005-06-23)",
/* 178 */ "???",
/* 179 */ "???",
/* 180 */ "???",
/* 181 */ "???",
/* 182 */ "???",
/* 183 */ "???",
/* 184 */ "???",
/* 185 */ "???",
/* 186 */ "???",
/* 187 */ "???",
/* 188 */ "???",
/* 189 */ "???",
/* 190 */ "???",
/* 191 */ "???",
/* 192 */ "???",
/* 193 */ "???",
/* 194 */ "???",
/* 195 */ "???",
/* 196 */ "???",
/* 197 */ "???",
/* 198 */ "???",
/* 199 */ "???",
/* 200 */ "???",
/* 201 */ "???",
/* 202 */ "???",
/* 203 */ "???",
/* 204 */ "???",
/* 205 */ "???",
/* 206 */ "???",
/* 207 */ "???",
/* 208 */ "PXELINUX Magic",
/* 209 */ "Configuration File",
/* 210 */ "Path Prefix",
/* 211 */ "Reboot Time",
/* 212 */ "OPTION_6RD",
/* 213 */ "OPTION_V4_ACCESS_DOMAIN",
/* 214 */ "???",
/* 215 */ "???",
/* 216 */ "???",
/* 217 */ "???",
/* 218 */ "???",
/* 219 */ "???",
/* 220 */ "Subnet Allocation",
/* 221 */ "Virtual Subnet Selection",
/* 222 */ "???",
/* 223 */ "???",
/* 224 */ "Reserved for private use",
/* 225 */ "Reserved for private use",
/* 226 */ "Reserved for private use",
/* 227 */ "Reserved for private use",
/* 228 */ "Reserved for private use",
/* 229 */ "Reserved for private use",
/* 230 */ "Reserved for private use",
/* 231 */ "Reserved for private use",
/* 232 */ "Reserved for private use",
/* 233 */ "Reserved for private use",
/* 234 */ "Reserved for private use",
/* 235 */ "Reserved for private use",
/* 236 */ "Reserved for private use",
/* 237 */ "Reserved for private use",
/* 238 */ "Reserved for private use",
/* 239 */ "Reserved for private use",
/* 240 */ "Reserved for private use",
/* 241 */ "Reserved for private use",
/* 242 */ "Reserved for private use",
/* 243 */ "Reserved for private use",
/* 244 */ "Reserved for private use",
/* 245 */ "Reserved for private use",
/* 246 */ "Reserved for private use",
/* 247 */ "Reserved for private use",
/* 248 */ "Reserved for private use",
/* 249 */ "MSFT - Classless route",
/* 250 */ "Reserved for private use",
/* 251 */ "Reserved for private use",
/* 252 */ "MSFT - WinSock Proxy Auto Detect",
/* 253 */ "Reserved for private use",
/* 254 */ "Reserved for private use",
/* 255 */ "End"};

const char *dhcp_message_types[] = {
/*   0 */ "undefined",
/*   1 */ "DHCPDISCOVER",
/*   2 */ "DHCPOFFER",
/*   3 */ "DHCPREQUEST",
/*   4 */ "DHCPDECLINE",
/*   5 */ "DHCPACK",
/*   6 */ "DHCPNAK",
/*   7 */ "DHCPRELEASE",
/*   8 */ "DHCPINFORM"
};

const char *netbios_node_type[] = {
/*   0 */ "none",
/*   1 */ "B-node",
/*   2 */ "P-node",
/*   3 */ "",
/*   4 */ "M-node",
/*   5 */ "",
/*   6 */ "",
/*   7 */ "",
/*   8 */ "H-node"
};

const char *option_overload[] = {
/*   0 */ "unspecified",
/*   1 */ "file field holds options",
/*   2 */ "sname field holds options",
/*   3 */ "file and sname field holds options"
};

const char *enabledisable[] = {
/*   0 */ "disabled",
/*   1 */ "enabled"
};

const char *ethernet_encapsulation[] = {
/*   0 */ "Ethernet version 2",
/*   1 */ "IEEE 802.3"
};

const char *operands[] = {
/*   0 */ "undefined",
/*   1 */ "BOOTPREQUEST",
/*   2 */ "BOOTPREPLY"
};

// From RFC 3046
const char *relayagent_suboptions[] = {
/*   0 */ "<unknown>",
/*   1 */ "Circuit-ID",
/*   2 */ "Remote-ID"
};

// Copied from RFC1700
const char *htypes[] = {
/*   0 */ "undefined",
/*   1 */ "Ethernet",
/*   2 */ "Experimental Ethernet",
/*   3 */ "Amateur Radio AX.25",
/*   4 */ "Proteon ProNET Token Ring",
/*   5 */ "Chaos",
/*   6 */ "IEEE 802 Networks",
/*   7 */ "ARCNET",
/*   8 */ "Hyperchannel",
/*   9 */ "Lanstar",
/*  10 */ "Autonet Short Address",
/*  11 */ "LocalTalk",
/*  12 */ "LocalNet",
/*  13 */ "Ultra link",
/*  14 */ "SMDS",
/*  15 */ "Frame Relay",
/*  16 */ "ATM",
/*  17 */ "HDLC",
/*  18 */ "Fibre Channel",
/*  19 */ "ATM",
/*  20 */ "Serial Line",
/*  21 */ "ATM"
};
